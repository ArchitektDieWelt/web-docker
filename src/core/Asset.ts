export type AssetPosition = "head" | "body";

export type AssetBase = {
  src: string;
  name?: string;
  position?: AssetPosition;
  priority?: number;
  countries?: string[];
};

export type AssetScript = AssetBase & {
  type: "js";
  async?: boolean;
  buildType?: "modern" | "legacy";
  deferred?: boolean;
};

export type AssetLink = AssetBase & {
  type: "css";
  media?: string;
};

export type Asset = AssetLink | AssetScript;
