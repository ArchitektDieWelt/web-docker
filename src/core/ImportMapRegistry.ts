interface ImportMap {
  [key: string]: string;
}

export class ImportMapRegistry {
  private importMap: ImportMap = {};

  constructor(
    private readonly script: HTMLScriptElement = document.createElement("script"),
    private readonly documentHead = document.head
  ) {
    this.script.type = "importmap";
    this.documentHead.appendChild(this.script);
  }

  public add(importMap: ImportMap) {
    this.importMap = {
      ...this.importMap,
      ...importMap,
    };
    this.script.textContent = JSON.stringify(this.importMap);
  }

  public get(): ImportMap {
    return this.importMap;
  }
}
