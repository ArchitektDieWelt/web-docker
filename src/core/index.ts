import RegisterEvent, { RegisterEventType } from "~/core/RegisterEvent";
import InitializeEvent, { InitializeEventType } from "~/core/InitializeEvent";

export { default as initialize } from "./initialize";

declare global {
  interface WindowEventMap {
    [RegisterEventType]: RegisterEvent;
    [InitializeEventType]: InitializeEvent;
  }
  interface Window {
    openreply: {
      initialize: (config: {
        configFilePath: string;
      }) => void;
    };
  }
}
