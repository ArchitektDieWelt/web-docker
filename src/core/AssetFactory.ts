import { Asset, AssetLink, AssetPosition, AssetScript } from "~/core/Asset";

export default class AssetFactory {
  private static buildTagScript(attr: AssetScript): HTMLScriptElement {
    const script = document.createElement("script");
    script.src = attr.src;
    if (attr.buildType === "modern") {
      script.type = "module";
      script.setAttribute("crossorigin", "");
    } else {
      script.type = "text/javascript";
      if (attr.async) script.setAttribute("async", "");
    }

    if (attr.priority)
      script.setAttribute("data-priority", attr.priority.toString());

    return script;
  }

  private static buildTagLink(attr: AssetLink): HTMLLinkElement {
    const link = document.createElement("link");
    link.setAttribute("rel", "stylesheet");
    link.setAttribute("href", attr.src);
    if (attr.media) link.setAttribute("media", attr.media);

    if (attr.priority)
      link.setAttribute("data-priority", attr.priority.toString());

    return link;
  }

  public create(
    assets: Asset[],
    position: AssetPosition = "head",
    countryCode?: string
  ): (HTMLLinkElement | HTMLScriptElement)[] {
    return (
      countryCode
        ? assets.filter(
            (asset) =>
              typeof asset.countries !== "object" ||
              asset.countries.includes(countryCode)
          )
        : assets
    )
      .filter((asset) => (asset.position ?? "head") === position)
      .sort((a, b) => {
        if (!a.priority || !b.priority) return 0;
        if (a.priority < b.priority) {
          return 1;
        } else if (a.priority > b.priority) {
          return -1;
        }
        return 0;
      })
      .map((entry) => {
        switch (entry.type) {
          case "css":
            return AssetFactory.buildTagLink(entry);
          case "js":
            return AssetFactory.buildTagScript(entry);
        }
      })
      .filter((item): item is HTMLLinkElement | HTMLScriptElement => !!item);
  }
}
