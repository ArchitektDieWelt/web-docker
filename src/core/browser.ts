import { InitializeEventType } from "~/core/InitializeEvent";
import initialize, { WebDockerOptions } from "~/core/initialize";

window.webdocker = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<WebDockerOptions>) => {
    await initialize(event.detail);
  }
);

