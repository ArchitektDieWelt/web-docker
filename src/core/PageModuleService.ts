import AssetFactory from "./AssetFactory";
import { Logger } from "~/core/Logger";
import type {
  IncludeType,
  PageInclude,
  PageModuleConfig,
} from "~/core/ModuleConfig";
import { ModuleService } from "~/core/ModuleService";

class PageModuleService implements ModuleService {
  private readonly logger: Logger = new Logger("PageModuleService");

  constructor(
    private readonly config: PageModuleConfig,
    private readonly assetFactory = new AssetFactory(),
    private readonly documentBody = document.body,
    private readonly documentHead = document.head
  ) {
    this.assetFactory = assetFactory;

    if (config.pages.length === 0 || config.pages.some(this.matches))
      this.injectAssets();
  }

  matches = (page: string | PageInclude): boolean => {
    if (this.isPageIncludeSemantics(page)) {
      console.warn(
        "PageInclude semantics are not implemented yet, please use a RegExp instead"
      );
      return false;
    } else {
      return !!window.location.pathname.match(new RegExp(page));
    }
  };

  isPageIncludeSemantics = (
    page: string | PageInclude
  ): page is PageInclude => {
    return (page as PageInclude).include !== undefined;
  };

  private injectAssets(): void {
    const headAssets = this.assetFactory.create(this.config.assets, "head");
    const bodyAssets = this.assetFactory.create(this.config.assets, "body");
    headAssets.forEach((asset) => this.documentHead.appendChild(asset));
    bodyAssets.forEach((asset) => this.documentBody.appendChild(asset));

    this.logger.log("injected assets in head", headAssets);
    this.logger.log("injected assets in body", bodyAssets);
  }

  get assetSources(): string[] {
    return this.config.assets.map((asset) => asset.src);
  }

  get module(): string {
    return this.config.module;
  }

  get type(): IncludeType {
    return this.config.type;
  }

  remove(): void {
    this.logger.warn("remove() is not implemented for PageModuleService");
  }
}
export { PageModuleService };
