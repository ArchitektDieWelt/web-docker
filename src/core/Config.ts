import { ModuleConfig } from "~/core/ModuleConfig";

export type Config = ModuleConfig;
