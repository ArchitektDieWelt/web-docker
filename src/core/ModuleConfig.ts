import { Asset } from "~/core/Asset";
import { Config } from "~/core/Config";

export type IncludeType = "observed" | "page";

export interface ModuleConfigBase {
  version: string;
  assets: Asset[];
  module: string;
  type: IncludeType;
}

export interface ObservedModuleConfig extends ModuleConfigBase {
  type: "observed";
  selector: string;
}

export interface PageInclude {
  include: string;
  exclude?: string;
  exact?: boolean;
}

export interface PageModuleConfig extends ModuleConfigBase {
  type: "page";
  pages: (string | PageInclude)[];
}

export type ModuleConfig = ObservedModuleConfig | PageModuleConfig;

export class ModuleConfigService {
  public getModuleConfig(config: Config): ModuleConfig {
    if (!config.version)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the version field`
      );

    if (!config.assets)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the assets field`
      );

    if (!config.module)
      throw Error(
        `Your tried to register a config with new structure:  ${JSON.stringify(
          config
        )}  it was missing the module field`
      );

    if (!config.type)
      throw Error(
        `Your tried to register a config with new structure: ${JSON.stringify(
          config
        )}  it was missing the type field`
      );

    if (config.type === "page" && !config.pages)
      throw Error(
        `Your tried to register a config with new structure of page include type:  ${JSON.stringify(
          config
        )}  it was missing the pages field`
      );

    return config;
  }
}
