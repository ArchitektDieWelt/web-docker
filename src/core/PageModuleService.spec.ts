import { describe, it, expect, beforeEach } from "vitest";
import { Asset } from "~/core/Asset";
import AssetFactory from "~/core/AssetFactory";

import { PageModuleConfig } from "~/core/ModuleConfig";
import { PageModuleService } from "~/core/PageModuleService";

const asset: Asset = {
  async: false,
  name: "chunk1.js",
  position: undefined,
  priority: 1,
  src: "http://src/1",
  deferred: false,
  type: "js",
};

describe("PageModuleService", () => {
  beforeEach(() => {
    document.body.innerHTML = "";
  });
  it("constructs", () => {
    window.location.href = "http://localhost/match";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: [".*"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };
    const service = new PageModuleService(config, assetFactoryMock);

    expect(service).toBeTruthy();
  });

  it("injects assets if URL of the page matches", async () => {
    window.location.href = "http://localhost/match";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: [".*"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });

  it("injects assets without custom element if URL of the page matches", async () => {
    window.location.href = "http://localhost/match";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: [".*"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });

  it("injects assets if empty page arrays was provided", async () => {
    window.location.href = "http://localhost/match";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: [],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });

  it("does not inject assets if URL of the page does not match", async () => {
    window.location.href = "http://localhost/match";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: ["does-not-match"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });

  it("injects assets if URL of the page matches the given regex pattern - as string", async () => {
    window.location.href = "http://localhost/matches/page";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: ["/matches/page"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });

  it("injects assets if URL of the page matches subpath of the given regex pattern - as string", async () => {
    window.location.href = "http://localhost/matches/page";
    const config: PageModuleConfig = {
      assets: [asset],
      type: "page",
      module: "test-module",
      pages: ["/matches"],
      version: "",
    };
    const assetFactoryMock: AssetFactory = {
      create(): (HTMLLinkElement | HTMLScriptElement)[] {
        return [document.createElement("link")];
      },
    };

    const service = new PageModuleService(config, assetFactoryMock);

    await service.load();

    expect(document.body).toMatchSnapshot();
  });
});
