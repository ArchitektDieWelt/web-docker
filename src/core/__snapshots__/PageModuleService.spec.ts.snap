// Vitest Snapshot v1

exports[`PageModuleService > does not inject assets if URL of the page does not match 1`] = `<body />`;

exports[`PageModuleService > injects assets if URL of the page matches 1`] = `
<body>
  <link />
</body>
`;

exports[`PageModuleService > injects assets if URL of the page matches subpath of the given regex pattern - as string 1`] = `
<body>
  <link />
</body>
`;

exports[`PageModuleService > injects assets if URL of the page matches the given regex pattern - as string 1`] = `
<body>
  <link />
</body>
`;

exports[`PageModuleService > injects assets if empty page arrays was provided 1`] = `
<body>
  <link />
</body>
`;

exports[`PageModuleService > injects assets without custom element if URL of the page matches 1`] = `
<body>
  <link />
</body>
`;
